### ISSUE TITLE
False positive PH2 error

### ISSUE THREAD

**Issue by TLWebdesign** [joomla-extensions/jedchecker#181]
Hi,

I have checked an extension and it got a false positive on the PH2 error.

It contained this code as the start of the file:

```
<?php
<?php
/**
 * @author ****
 * @copyright 2016
 * @version  2.0 (10-01-2022)
 * @license   <a href="http://www.gnu.org/licenses/gpl-3.0.html" target="_blank">GNU/GPLv3</a>
 * 
 */
defined("_JEXEC") or die("Restricted access");

```

It caught the error as in that it got double php opening tags but it listed it as  PH2 error.

Kind regards,

Tom

**Comment by Llewellynvdm** [comment]
Okay this makes sense and yes doing some PHP syntax validation would be helpful, what approach do you have in mind?

What we also need, to detect depreciated methods, classes and method signature mismatching of the Joomla API, this will a be a huge step-up... but something I would really want to help happen.

**Comment by dryabov** [comment]
> what approach do you have in mind?

https://github.com/nikic/PHP-Parser, but it requires PHP7.0 (currently required PHP version in JED Checker is 5.6).

> What we also need, to detect depreciated methods, classes and method signature mismatching of the Joomla API

It's quite complicated, because of dynamical nature of PHP, so any code analyzer cannot be sure about the type of a given variable. Probably, the best solution here is to use https://github.com/ircmaxell/php-cfg to track variables lifetimes, but it has some known issues (doesn't take into account parameters passed by reference and doesn't process try/catch/finally blocks properly).

An alternative is to use PHPStan with https://github.com/phpstan/phpstan-deprecation-rules, but it's quite a large project (it's shipped as a 20Mb phpstan.phar file). And I'm not sure it is able to cover nontrivial cases.
